### PR TITLE
Test Coverage PR-1: Tests added in ./licensecheck and ./repodata subdir

### PR DESCRIPTION
**Pull request by mohanmanikanta2299**
## Summary
This PR is the 1st part of the test coverage PRs intended to increase the test coverage of the entire repo from 55% to 80+%, fulfilling our coverage requirements.

## Changes Made
* Implemented comprehensive table-driven tests for the `./licensecheck` and `./repodata` packages.
* Added validation for CLI command flags, ensuring default values and flag bindings are properly configured.
* Added strict assertions (`assert.Equal`, `assert.Contains`) and explicit error checking to validate function outputs and edge cases.
* Used `require` for critical initializations (like loggers and flag lookups) to prevent nil-pointer panics during test execution.

## Testing
* Ran `go test -v ./...` (All tests passing)
* Ran `go test -race ./...` (No race conditions detected)
* Verified coverage metrics locally.

### :link: External Links
https://hashicorp.atlassian.net/browse/CCEN-512

### :+1: Definition of Done
- Tests added : ✅

### :thinking: Can be merged upon approval?
:white_check_mark:

## PCI review checklist

- [ ] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

### CODE REVIEW

File: licensecheck/licensecheck_test.go [hunk 1]
```diff
@@ -85,14 +85,137 @@ func TestEnsureCorrectName(t *testing.T) {
...
+		assert.Nil(t, err)
+
+		// Read file and verify header was prepended
+		content, _ := afero.ReadFile(AppFs, filePath)
+		assert.Contains(t, string(content), header)
+		assert.Contains(t, string(content), originalContent)
+		// Header should come before original content
+		headerIdx := lo.IndexOf([]byte(string(content)), []byte(header)[0])
+		contentIdx := lo.IndexOf([]byte(string(content)), []byte(originalContent)[0])
+		assert.Less(t, headerIdx, contentIdx)
```
Comment (CreatorHead) [inline licensecheck/licensecheck_test.go]: Image

Comment (CreatorHead) [reply licensecheck/licensecheck_test.go]: Hey @mohanmanikanta2299, I changed header to just "C" in the "add header to file with existing content" sub-test and the test still passed. Please take a look and review the PR properly before sending it for review.

```
aatish@aatish-CLPJCQ9LWW copywrite % go test ./licensecheck/ -run "TestAddHeader/add_header_to_file_with_existing_content" -v -count=1
=== RUN   TestAddHeader
=== RUN   TestAddHeader/add_header_to_file_with_existing_content
--- PASS: TestAddHeader (0.00s)
    --- PASS: TestAddHeader/add_header_to_file_with_existing_content (0.00s)
PASS
ok      github.com/hashicorp/copywrite/licensecheck     0.971s
```

**Review by CreatorHead — CHANGES_REQUESTED** [review]

File: licensecheck/licensecheck_test.go (hunk 1 above)
Comment (CreatorHead) [reply licensecheck/licensecheck_test.go]: @mohanmanikanta2299 , please enable co-pilot review as well, it will help saving some time in review from other Devs.

Comment (mohanmanikanta2299) [reply licensecheck/licensecheck_test.go]: Addressed the concern. 
Thanks

File: licensecheck/copyright_test.go [hunk 2]
```diff
@@ -273,3 +274,91 @@ func TestHasCopyright(t *testing.T) {
...
+
+	t.Run("file larger than 300 bytes with copyright after header", func(t *testing.T) {
+		// Create content > 300 bytes with copyright appearing after byte 300
+		header := make([]byte, 350)
+		for i := range header {
+			header[i] = 'X'
+		}
+		content := string(header) + "\nCopyright notice here"
+
+		f, _ := afero.TempFile(AppFs, tempDir, "")
```
Comment (CreatorHead) [inline licensecheck/copyright_test.go]: Several subtests silently discard errors with _, inconsistent with `require.NoError` used elsewhere:

If `afero.TempFile` fails, `f` is nil, and `f.Name()` panics. These should use `require.NoError`.

File: licensecheck/copyright_test.go [hunk 3]
```diff
@@ -273,3 +274,91 @@ func TestHasCopyright(t *testing.T) {
...
+		_ = afero.WriteFile(AppFs, f.Name(), []byte(content), 0644)
```
Comment (CreatorHead) [inline licensecheck/copyright_test.go]: For consistency we should use `require.NoError`.

File: licensecheck/licensecheck_test.go [hunk 4]
```diff
@@ -85,14 +87,138 @@ func TestEnsureCorrectName(t *testing.T) {
...
+		tempDir := t.TempDir()
+		filePath := filepath.Join(tempDir, "test.txt")
+		_ = afero.WriteFile(AppFs, filePath, []byte(""), 0644)
+
+		header := "Copyright (c) 2023 Test Corp"
+		err := AddHeader(filePath, header)
+		assert.Nil(t, err)
+
+		// Read file and verify header was added
+		content, _ := afero.ReadFile(AppFs, filePath)
```
Comment (CreatorHead) [inline licensecheck/licensecheck_test.go]: For consistency we should use `require.NoError`?

File: licensecheck/licensecheck_test.go [hunk 5]
```diff
@@ -85,14 +87,138 @@ func TestEnsureCorrectName(t *testing.T) {
...
+		originalContent := "This is the original file content"
+		err := afero.WriteFile(AppFs, filePath, []byte(originalContent), 0644)
+		require.NoError(t, err)
+
+		header := "Copyright (c) 2023 Test Corp"
+		err = AddHeader(filePath, header)
+		assert.Nil(t, err)
+
+		// Read file and verify header was prepended
+		content, _ := afero.ReadFile(AppFs, filePath)
```
Comment (CreatorHead) [inline licensecheck/licensecheck_test.go]: For consistency we should use `require.NoError`

File: licensecheck/licensecheck_test.go [hunk 6]
```diff
@@ -85,14 +87,138 @@ func TestEnsureCorrectName(t *testing.T) {
...
+	t.Run("add multi-line header", func(t *testing.T) {
+		tempDir := t.TempDir()
+		filePath := filepath.Join(tempDir, "test.txt")
+		_ = afero.WriteFile(AppFs, filePath, []byte("Original content"), 0644)
+
+		header := "Copyright (c) 2023 Test Corp\nSPDX-License-Identifier: MPL-2.0"
+		err := AddHeader(filePath, header)
+		assert.Nil(t, err)
+
+		content, _ := afero.ReadFile(AppFs, filePath)
```
Comment (CreatorHead) [inline licensecheck/licensecheck_test.go]: same as above comments

File: licensecheck/update_test.go [hunk 7]
```diff
@@ -994,3 +995,259 @@ package main
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tempDir := t.TempDir()
+
+			fileName := "test.go"
+			if tt.name == "Skip .copywrite.hcl file" {
+				fileName = ".copywrite.hcl"
+			}
```
Comment (CreatorHead) [inline licensecheck/update_test.go]: Please check, I think this is an anti-pattern design? (CMIIW) :
The file name shouldn't be derived from `tt.name`, if the test name is ever reworded or typo'd, this silently breaks the test case it's meant to cover without any compile-time or runtime signal?

File: licensecheck/update_test.go [hunk 8]
```diff
@@ -994,3 +995,259 @@ package main
...
+	t.Run("executeGitCommand", func(t *testing.T) {
+		out, err := executeGitCommand(tempDir, "log", "-1", "--format=%s")
+		require.NoError(t, err)
+		assert.Equal(t, "second commit\n", string(out))
+	})
+
+	t.Run("InitializeGitCache", func(t *testing.T) {
+		evalRoot, _ := filepath.EvalSymlinks(tempDir)
+		once = sync.Once{}
+		err := InitializeGitCache(evalRoot)
```
Comment (CreatorHead) [inline licensecheck/update_test.go]: These subtests directly reset shared package variables like once, firstCommitYearCached, and lastCommitYearsCache. What happens if these tests ever run in parallel?

File: licensecheck/copyright_test.go [hunk 9]
```diff
@@ -273,3 +274,91 @@ func TestHasCopyright(t *testing.T) {
...
+		require.NoError(t, err)
+
+		hasCopyright, err := HasMatchingCopyright(f.Name(), "Copyright", false)
+		assert.Nil(t, err)
+		assert.True(t, hasCopyright)
+	})
+
+	t.Run("file larger than 300 bytes with copyright after header", func(t *testing.T) {
+		// Create content > 300 bytes with copyright appearing after byte 300
+		header := make([]byte, 350)
```
Comment (CreatorHead) [inline licensecheck/copyright_test.go]: Magic number embedded in test assertion?

File: licensecheck/update_test.go [hunk 10]
```diff
@@ -994,3 +995,259 @@ package main
...
+		require.NoError(t, err)
+		assert.Equal(t, 2020, firstCommitYearCached)
+		assert.Contains(t, lastCommitYearsCache, "test.txt")
+		assert.Contains(t, lastCommitYearsCache, "test2.txt")
+		assert.Equal(t, 2020, lastCommitYearsCache["test.txt"])
+		assert.Equal(t, 2023, lastCommitYearsCache["test2.txt"])
+	})
+
+	t.Run("InitializeGitCache - Failure", func(t *testing.T) {
```
Comment (CreatorHead) [inline licensecheck/update_test.go]: The test name says "Failure" but asserts `NoError`. it seems to be testing that the function swallows errors gracefully. If that's intentional, rename the subtest to reflect the expected behavior, and add a comment explaining why errors are suppressed here.

File: licensecheck/licensecheck_test.go (hunk 1 above)
Comment (CreatorHead) [reply licensecheck/licensecheck_test.go]: This is still not fixed — assert.Contains with "C" as the header still passes since "C" exists in the original content.

**Review by CreatorHead — CHANGES_REQUESTED** [review]

File: licensecheck/licensecheck_test.go (hunk 1 above)
Comment (mohanmanikanta2299) [reply licensecheck/licensecheck_test.go]: This unit test is written for the method [AddHeader](https://github.com/hashicorp/copywrite/blob/mohan/test_coverage/licensecheck/licensecheck.go#L37-L44) in `licensecheck.go` file. The purpose of this method is to read the existing the file content, prepend `header+"\n\n"` to the file and write it back into the file.

Since it is a unit test written and we are mocking the parameters `filepath` and `header`, any valid String value in `header` will be appended to the given valid `filepath` mocked in the unit test. 
Even if we write `"C"` or `"Copyright (c) 2023 Test Corp"`, the header should be prepended to the file since both of them are valid strings.

We are not validating the Copyright Header in the `AddHeader` method which we are testing.

However, I have tightened assertions around the test case to make sure the order of statements written are correct.

File: licensecheck/copyright_test.go (hunk 2 above)
Comment (mohanmanikanta2299) [reply licensecheck/copyright_test.go]: Used require.NoError and validated all errors instead of silently skipping them

File: licensecheck/update_test.go (hunk 7 above)
Comment (mohanmanikanta2299) [reply licensecheck/update_test.go]: Made changes and added fileName parameter in struct to get the file name instead of depending on the tt.Name to determine the fileName for testing.

File: licensecheck/copyright_test.go (hunk 9 above)
Comment (mohanmanikanta2299) [reply licensecheck/copyright_test.go]: Updated copyright.go file which has the magic number hardcoded and updated tests to avoid magic numbers and fetch variable `copyrightHeaderBytes`

File: licensecheck/update_test.go (hunk 10 above)
Comment (mohanmanikanta2299) [reply licensecheck/update_test.go]: Renamed the test case and added comments explaining the expected behaviour.

File: licensecheck/update_test.go (hunk 8 above)
Comment (mohanmanikanta2299) [reply licensecheck/update_test.go]: These subtests share package-level variables `once`, `lastCommitYearsCache`, `firstCommitYearCached`.
Each subtest needs to reset that cache to test a different scenario and that reset is not safe to do concurrently. `sync.Once` cannot be reassigned while another goroutine is inside [`once.Do`](https://github.com/hashicorp/copywrite/blob/mohan/test_coverage/licensecheck/update.go#L480), and the map has no lock protecting it.

In production this is fine because `InitializeGitCache` is called only one time before the worker pool starts, so by the time goroutines read the map it's already fully written. But in tests we're deliberately forcing re-initialisation.

Adding `t.Parallel()` here would cause data races that `go test -race` would catch. Sequential execution is the right call for this test. The `resetGitCacheForTest` helper and the comment on `TestGitOperations` are added to make that constraint visible so no one accidentally changes it later.

**Review by CreatorHead — APPROVED** [review]
